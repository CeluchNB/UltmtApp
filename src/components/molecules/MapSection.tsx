import * as React from 'react'
import { useColors } from '../../hooks'
import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
import { Button, IconButton } from 'react-native-paper'
import { size, weight } from '../../theme/fonts'

interface MapSectionProps<T> {
    title: string
    listData?: T[]
    renderItem: (item: T) => JSX.Element
    showButton: boolean
    showCreateButton: boolean
    loading?: boolean
    buttonText?: string
    onButtonPress?: () => void
    onCreatePress?: () => void
    error?: string
}

const MapSection = <T,>({
    title,
    listData,
    renderItem,
    showButton,
    showCreateButton,
    loading,
    buttonText,
    onButtonPress,
    onCreatePress,
    error,
}: MapSectionProps<T>) => {
    const { colors } = useColors()

    const styles = StyleSheet.create({
        container: {
            width: '100%',
        },
        titleContainer: {
            flex: 1,
            display: 'flex',
            flexDirection: 'row',
        },
        createButton: {
            display: 'flex',
            flexWrap: 'nowrap',
            backgroundColor: colors.textPrimary,
            borderRadius: 5,
        },
        title: {
            flex: 1,
            fontSize: size.fontLarge,
            color: colors.textPrimary,
            fontWeight: weight.bold,
            marginBottom: 10,
        },
        button: {
            alignSelf: 'center',
            marginTop: 10,
        },
        error: {
            fontSize: size.fontMedium,
            marginBottom: 50,
            color: colors.textPrimary,
        },
    })

    return (
        <View style={styles.container}>
            <View style={styles.titleContainer}>
                <Text style={styles.title}>{title}</Text>
                {showCreateButton && (
                    <IconButton
                        color={colors.primary}
                        style={styles.createButton}
                        onPress={onCreatePress}
                        icon="plus"
                        testID="create-button"
                    />
                )}
            </View>
            {error && <Text style={styles.error}>{error}</Text>}
            {!error &&
                !loading &&
                listData &&
                listData.map(item => renderItem(item))}
            {loading && (
                <ActivityIndicator
                    size="large"
                    color={colors.textPrimary}
                    animating={loading}
                />
            )}
            {showButton && (
                <Button
                    mode="text"
                    style={styles.button}
                    color={colors.textPrimary}
                    onPress={onButtonPress}
                    testID="more-button">
                    {buttonText}
                </Button>
            )}
        </View>
    )
}

export default MapSection
