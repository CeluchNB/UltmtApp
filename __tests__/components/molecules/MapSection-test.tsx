import '@testing-library/jest-native/extend-expect'
import * as React from 'react'
import MapSection from '../../../src/components/molecules/MapSection'
import { Text } from 'react-native'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react-native'

it('matches snapshot', () => {
    const snapshot = renderer.create(
        <MapSection
            title="title"
            listData={['item1', 'item2']}
            renderItem={item => {
                return <Text key={item}>{item}</Text>
            }}
            showButton={true}
            showCreateButton={true}
            loading={true}
        />,
    )

    expect(snapshot).toMatchSnapshot()
})

it('displays items correctly', () => {
    const { getByText } = render(
        <MapSection
            title="title"
            listData={['item1', 'item2']}
            renderItem={item => {
                return <Text key={item}>{item}</Text>
            }}
            showButton={true}
            showCreateButton={true}
        />,
    )

    const item1 = getByText('item1')
    const item2 = getByText('item2')
    expect(item1).toBeTruthy()
    expect(item2).toBeTruthy()
})

it('button press fires', () => {
    const fn = jest.fn()
    const { queryByTestId } = render(
        <MapSection
            title="title"
            listData={['item1', 'item2']}
            renderItem={item => {
                return <Text key={item}>{item}</Text>
            }}
            showButton={true}
            onButtonPress={fn}
            showCreateButton={true}
        />,
    )

    const button = queryByTestId('more-button')
    if (!button) {
        throw new Error('button does not exist')
    }
    fireEvent.press(button)
    expect(fn).toHaveBeenCalled()
})

it('create button fires', () => {
    const fn = jest.fn()
    const { queryByTestId } = render(
        <MapSection
            title="title"
            listData={['item1', 'item2']}
            renderItem={item => {
                return <Text key={item}>{item}</Text>
            }}
            showButton={true}
            showCreateButton={true}
            onCreatePress={fn}
        />,
    )

    const button = queryByTestId('create-button')
    if (!button) {
        throw new Error('button does not exist')
    }
    fireEvent.press(button)
    expect(fn).toHaveBeenCalled()
})
