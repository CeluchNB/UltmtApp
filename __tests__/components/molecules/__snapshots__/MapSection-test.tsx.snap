// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<View
  style={
    {
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffd600",
          "flex": 1,
          "fontSize": 61.64383561643835,
          "fontWeight": "bold",
          "marginBottom": 10,
        }
      }
    >
      title
    </Text>
    <View
      collapsable={false}
      style={
        {
          "alignSelf": undefined,
          "bottom": undefined,
          "left": undefined,
          "position": undefined,
          "right": undefined,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "top": undefined,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#ffd600",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 5,
              "borderWidth": 0,
              "display": "flex",
              "elevation": 0,
              "flexWrap": "nowrap",
              "height": 40,
              "margin": 6,
              "overflow": "hidden",
              "width": 40,
            }
          }
        >
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityState={
              {
                "disabled": true,
              }
            }
            accessibilityTraits="button"
            accessible={true}
            centered={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                false,
                [
                  {
                    "alignItems": "center",
                    "flexGrow": 1,
                    "justifyContent": "center",
                  },
                  {
                    "borderRadius": 5,
                  },
                ],
              ]
            }
            testID="create-button"
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                [
                  {
                    "color": "#303030",
                    "fontSize": 24,
                  },
                  [
                    {
                      "lineHeight": 24,
                      "transform": [
                        {
                          "scaleX": 1,
                        },
                      ],
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ],
                  {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              Û∞êï
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <ActivityIndicator
    animating={true}
    color="#ffd600"
    size="large"
  />
  <View
    collapsable={false}
    style={
      {
        "alignSelf": "center",
        "bottom": undefined,
        "left": undefined,
        "position": undefined,
        "right": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
        "top": undefined,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
    >
      <View
        collapsable={false}
        color="#ffd600"
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 0,
            "marginTop": 10,
            "minWidth": 64,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              false,
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="more-button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 12,
                    },
                    undefined,
                    false,
                    {
                      "color": "rgba(103, 80, 164, 1)",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    undefined,
                  ],
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
